core/src/main/java/org/apache/iceberg/actions/FileRewriteGroup.java

```diff
@@ -0,0 +1,104 @@
+
+/**
+ * Container class representing a set of files to be rewritten by a {@link FileRewriteExecutor}.
+ *
+ * @param <I> the Java type of the plan info member like {@link RewriteDataFiles.FileGroupInfo} or
+ *     {@link RewritePositionDeleteFiles.FileGroupInfo}
+ * @param <T> the Java type of the tasks to read the files which are rewritten
+ * @param <F> the Java type of the content files which are rewritten
+ */
+public abstract class FileRewriteGroup<I, T extends ContentScanTask<F>, F extends ContentFile<F>> {
+  private final I info;
+  private final List<T> fileScanTasks;
+  private final long writeMaxFileSize;
+  private final long splitSize;
+  private final int expectedOutputFiles;
+
+  FileRewriteGroup(
+      I info,
+      List<T> fileScanTasks,
+      long writeMaxFileSize,
+      long splitSize,
+      int expectedOutputFiles) {
+    this.info = info;
+    this.fileScanTasks = fileScanTasks;
+    this.writeMaxFileSize = writeMaxFileSize;
+    this.splitSize = splitSize;
+    this.expectedOutputFiles = expectedOutputFiles;
+  }
+
+  /** Identifiers and partition information about the group. */
+  public I info() {
+    return info;
+  }
+
+  /** Input of the group. {@link ContentScanTask}s to read. */
+  public List<T> fileScans() {
+    return fileScanTasks;
+  }
+
+  /**
+   * While we create tasks that should all be smaller than our target size, there is a chance that
+   * the actual data will end up being larger than our target size due to various factors of
+   * compression, serialization, which are outside our control. If this occurs, instead of making a
+   * single file that is close in size to our target, we would end up producing one file of the
+   * target size, and then a small extra file with the remaining data.
+   *
+   * <p>For example, if our target is 512 MB, we may generate a rewrite task that should be 500 MB.
+   * When we write the data we may find we actually have to write out 530 MB. If we use the target
+   * size while writing, we would produce a 512 MB file and an 18 MB file. If instead we use a
+   * larger size estimated by this method, then we end up writing a single file.
+   *
+   * @return the target size plus one half of the distance between max and target
+   */
+  public long writeMaxFileSize() {
+    return writeMaxFileSize;
+  }
+
+  /**
+   * Determines the reader split size as the input size divided by the desired number of output
+   * files. The final split size is adjusted to be at least as big as the target file size but less
+   * than the max write file size.
+   */
+  public long splitSize() {
+    return splitSize;
+  }
+
+  /** Expected number of the output files. */
+  public int expectedOutputFiles() {
+    return expectedOutputFiles;
+  }
+
+  /** Accumulated size for the input files. */
+  public long sizeInBytes() {
+    return fileScanTasks.stream().mapToLong(T::length).sum();
+  }
+
+  /** Number of the input files. */
+  public int numFiles() {
```

Comment:
nit: `numInputFiles`

Reply:
This method exists in the current API as `RewriteFileGroup.numFiles()` and `RewritePositionDeletesGroup.numRewrittenDeleteFiles()` and the new method name will be inherited from the FileRewriteGroup

Minimally one of them needs to be deprecated, renamed. Do you think we should change both?